### PR TITLE
add the transpose op

### LINKED ISSUE #4163: Transpose Operator
add transpose op
this op likes an python numpy transpose function
which 
array1 = np.random.randn(1, 2, 3, 4)
array2 = array1.transpose((0, 2, 3, 1))

and the shape of array2 is (1, 3, 4, 2).   The order of data will also change

### PR DESCRIPTION
**Pull request by NHZlX**
fix #4163

### CODE REVIEW

File: paddle/operators/transpose_op.cu [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+  auto buffer_dims = framework::make_ddim(buffer_dim_shape);
+  framework::Tensor host_buffer;
+  platform::CPUPlace cpu_place;
+  platform::GPUPlace gpu_place;
+
+  int* host_buffer_data = host_buffer.mutable_data<int>(buffer_dims, cpu_place);
+
+  auto offset_buffer =
+      memory::Alloc(context.GetPlace(), ndims * 3 * sizeof(int));
```
Comment (QiJune) [inline paddle/operators/transpose_op.cu]: 这里接口使用不当，context.GetPlace() 返回的是 Place，是一个boost::variant<CPUPlace, GPUPlace>，而Alloc需要接收一个确定的GPUPlace or CPUPlace

File: paddle/operators/transpose_op.cu [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  }
+
+  for (int i = 0; i < ndims; i++) {
+    host_buffer_data[i] = in_offset[i];
+    host_buffer_data[i + ndims] = out_offset[i];
+    host_buffer_data[i + ndims * 2] = axis[i];
+  }
+
+  memory::Copy(gpu_place, offset_buffer, cpu_place, host_buffer_data,
+               ndims * 3 * sizeof(int));
```
Comment (QiJune) [inline paddle/operators/transpose_op.cu]: 在memory copy的时候，需要指定对应的cuda Stream

```
template <typename DstPlace, typename SrcPlace>
void Copy(DstPlace, void* dst, SrcPlace, const void* src, size_t num,
          cudaStream_t stream);
```

cuda stream 可以从context里面拿

File: paddle/operators/transpose_op.cu (hunk 1 above)
Comment (NHZlX) [reply paddle/operators/transpose_op.cu]: Done

File: paddle/operators/transpose_op.cu (hunk 2 above)
Comment (NHZlX) [reply paddle/operators/transpose_op.cu]: Done

File: paddle/operators/transpose_op.h [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+      case 3:
+        DoTranspose<Place, T, 3>(context, *in, *out, axis);
+        break;
+      case 4:
+        DoTranspose<Place, T, 4>(context, *in, *out, axis);
+        break;
+      case 5:
+        DoTranspose<Place, T, 5>(context, *in, *out, axis);
+        break;
+      default:
```
[the remark this answers is not in the document]
Comment (NHZlX) [reply paddle/operators/transpose_op.h]: When ndmis > 5, calling NaiveCpuTranspose

File: paddle/operators/transpose_op.h [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    for (size_t i = 0; i < ndims; i++) {
+      from_index += (temp / out_offset[i]) * in_offset[axis[i]];
+      temp = temp % out_offset[i];
+    }
+    out_data[to_index] = in_data[from_index];
+  }
+}
+
+template <typename Place, typename T, int Dims>
+void DoTranspose(const framework::ExecutionContext& context,
```
[the remark this answers is not in the document]
Comment (NHZlX) [reply paddle/operators/transpose_op.h]: sounds great

File: paddle/operators/transpose_op.h [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        break;
+      case 3:
+        DoTranspose<Place, T, 3>(context, *in, *out, axis);
+        break;
+      case 4:
+        DoTranspose<Place, T, 4>(context, *in, *out, axis);
+        break;
+      case 5:
+        DoTranspose<Place, T, 5>(context, *in, *out, axis);
+        break;
```
[the remark this answers is not in the document]
Comment (NHZlX) [reply paddle/operators/transpose_op.h]: ok，如果对与维度大于5的不支持的话，是可以的，省了很多代码，包括NaiveCpuTranspose 以及 Gpu kernel的代码

File: paddle/operators/transpose_op.h [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+template <typename Place, typename T>
+class TransposeGradKernel : public framework::OpKernel {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto* in = context.Input<framework::Tensor>(framework::GradVarName("Out"));
+    auto* out = context.Output<framework::Tensor>(framework::GradVarName("X"));
+    out->mutable_data<T>(context.GetPlace());
+
+    auto axis_temp = context.Attr<std::vector<int>>("axis");
+    std::vector<int> axis(axis_temp);
```
[the remark this answers is not in the document]
Comment (NHZlX) [reply paddle/operators/transpose_op.h]: there is no grad for axis

File: paddle/operators/transpose_op.cc [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+      PADDLE_ENFORCE_EQ(axis_sorted[i], (int)i,
+                        "the sorted axis should be [0, 1, ... dims - 1], "
+                        "the dims equals to the input tensor dimensions");
+    }
+
+    framework::DDim out_dim(in_dim);
+    for (size_t i = 0; i < axis.size(); i++) {
+      out_dim[i] = in_dim[axis[i]];
+    }
+    ctx.Output<Tensor>("Out")->Resize(out_dim);
```
[the remark this answers is not in the document]
Comment (NHZlX) [reply paddle/operators/transpose_op.cc]: Done

File: paddle/operators/transpose_op.cc [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("X"), "Input(X) should not be null");
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar(framework::GradVarName("Out")),
+                            "Input(Out@GRAD) should not be null");
+    auto x_dims = ctx.Input<Tensor>("X")->dims();
+    auto *x_grad = ctx.Output<Tensor>(framework::GradVarName("X"));
```
[the remark this answers is not in the document]
Comment (NHZlX) [reply paddle/operators/transpose_op.cc]: Done

File: paddle/operators/transpose_op.cu [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+#include "paddle/memory/memcpy.h"
+#include "paddle/memory/memory.h"
+#include "paddle/operators/transpose_op.h"
+
+namespace paddle {
+namespace operators {
+
+template <typename T>
+__global__ void transpose_kernel(int nthreads, const T* in_data, T* out_data,
```
[the remark this answers is not in the document]
Comment (NHZlX) [reply paddle/operators/transpose_op.cu]: have deleted this kernel and used the eigen interface

File: python/paddle/v2/framework/tests/test_transpose_op.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import unittest
+import numpy as np
+from gradient_checker import GradientChecker
+from op_test_util import OpTestMeta
+from paddle.v2.framework.op import Operator
```
[the remark this answers is not in the document]
Comment (NHZlX) [reply python/paddle/v2/framework/tests/test_transpose_op.py]: Done

File: python/paddle/v2/framework/tests/test_transpose_op.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+
+
+class TransposeGradOpTest(GradientChecker):
+    def test_transpose(self):
+        op = Operator("transpose", X="X", Out="Out", axis=[1, 0])
+        inputs = {'X': np.random.random((32, 84)).astype("float32"), }
+
+        self.check_grad(op, inputs, set(["X"]), "Out", max_relative_error=0.5)
+
+
```
[the remark this answers is not in the document]
Comment (NHZlX) [reply python/paddle/v2/framework/tests/test_transpose_op.py]: Done

File: paddle/operators/transpose_op.cc [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+using framework::Tensor;
+
+class TransposeOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("Input"),
+                            "Input(Input) should not be null");
```
Comment (Xreki) [inline paddle/operators/transpose_op.cc]: 对输出也需要检查

File: paddle/operators/transpose_op.cc [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+};
+
+class TransposeOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  TransposeOpMaker(framework::OpProto *proto,
+                   framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput(
+        "Input",
+        "(Tensor)The input tensor, tensors with rank at most 7 are supported");
```
Comment (Xreki) [inline paddle/operators/transpose_op.cc]: 最高只支持6-D吧？

File: paddle/operators/transpose_op.cc [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+                            "Input(Output@GRAD) should not be null");
+    auto input_dims = ctx.Input<Tensor>("Input")->dims();
+    auto *input_grad =
+        ctx.Output<framework::LoDTensor>(framework::GradVarName("Input"));
+
+    auto output_grad_dims =
+        ctx.Input<Tensor>(framework::GradVarName("Output"))->dims();
+    auto output_dims = ctx.Input<Tensor>("Output")->dims();
+
+    PADDLE_ENFORCE(output_grad_dims == output_dims,
```
Comment (Xreki) [inline paddle/operators/transpose_op.cc]: 可以使用`PADDLE_ENFORCE_EQ`

File: paddle/operators/transpose_op.h [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+    for (size_t i = 0; i < axis.size(); i++) {
+      axis[axis_temp[i]] = i;
+    }
+
+    int ndims = axis.size();
+
+    switch (ndims) {
+      case 1:
+        break;
```
Comment (Xreki) [inline paddle/operators/transpose_op.h]: 同上

File: python/paddle/v2/framework/tests/test_transpose_op.py [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+class TestCase3(TestTransposeOp):
+    def initTestCase(self):
+        self.shape = (2, 3, 4, 5, 6)
+        self.axis = (4, 2, 3, 1, 0)
+
+
+class TestCase4(TestTransposeOp):
+    def initTestCase(self):
+        self.shape = (2, 3, 4, 5, 6, 1)
+        self.axis = (4, 2, 3, 1, 0, 5)
```
Comment (Canpio) [inline python/paddle/v2/framework/tests/test_transpose_op.py]: Here should be a test about whether our Op can throw an exception correctly. However, our framework is not able to support such a test right now. So I leave a comment here to remind us there is something TODO. I have created an issue about this: https://github.com/PaddlePaddle/Paddle/issues/4173

File: paddle/operators/transpose_op.cc (hunk 13 above)
Comment (NHZlX) [reply paddle/operators/transpose_op.cc]: 对， 6D以上不支持

File: paddle/operators/transpose_op.h [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto* input = context.Input<framework::Tensor>("Input");
+    auto* output = context.Output<framework::Tensor>("Output");
+    output->mutable_data<T>(context.GetPlace());
+
+    auto axis = context.Attr<std::vector<int>>("axis");
+    int ndims = axis.size();
+    switch (ndims) {
+      case 1:
```
[the remark this answers is not in the document]
Comment (NHZlX) [reply paddle/operators/transpose_op.h]: 这是个bug， 已修复

File: paddle/operators/transpose_op.h (hunk 15 above)
Comment (NHZlX) [reply paddle/operators/transpose_op.h]: Done

File: paddle/operators/transpose_op.cc [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+class TransposeOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("Input"),
+                            "Input(Input) should not be null");
+    auto input_dim = ctx.Input<Tensor>("Input")->dims();
+    auto axis = ctx.Attr<std::vector<int>>("axis");
```
[the remark this answers is not in the document]
Comment (NHZlX) [reply paddle/operators/transpose_op.cc]: Done

File: paddle/operators/transpose_op.cc [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+    size_t input_dim_size = input_dim.size();
```
[the remark this answers is not in the document]
Comment (NHZlX) [reply paddle/operators/transpose_op.cc]: Done

File: paddle/operators/transpose_op.cc [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+                      "should be equal to the axis's size(%d)",
+                      input_dim_size, axis_size);
+
+    std::vector<int> axis_sorted(axis);
+    std::sort(axis_sorted.begin(), axis_sorted.end());
+    for (size_t i = 0; i < axis_sorted.size(); i++) {
+      PADDLE_ENFORCE_EQ(axis_sorted[i], static_cast<int>(i),
+                        "the sorted axis should be [0, 1, ... dims - 1], "
+                        "where the dims is the axis's size");
+    }
```
[the remark this answers is not in the document]
Comment (NHZlX) [reply paddle/operators/transpose_op.cc]: Done

File: paddle/operators/transpose_op.h [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+
+template <typename Place, typename T>
+class TransposeGradKernel : public framework::OpKernel {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto* output_grad =
+        context.Input<framework::Tensor>(framework::GradVarName("Output"));
+    auto* input_grad =
+        context.Output<framework::Tensor>(framework::GradVarName("Input"));
+    input_grad->mutable_data<T>(context.GetPlace());
```
[the remark this answers is not in the document]
Comment (NHZlX) [reply paddle/operators/transpose_op.h]: Done

File: paddle/operators/transpose_op.h [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+
+    auto axis_temp = context.Attr<std::vector<int>>("axis");
+    std::vector<int> axis(axis_temp);
+
+    for (size_t i = 0; i < axis.size(); i++) {
+      axis[axis_temp[i]] = i;
```
[the remark this answers is not in the document]
Comment (NHZlX) [reply paddle/operators/transpose_op.h]: good choice

File: paddle/operators/transpose_op.cc [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+
+class TransposeOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  TransposeOpMaker(framework::OpProto *proto,
+                   framework::OpAttrChecker *op_checker)
+      : OpProtoAndCheckerMaker(proto, op_checker) {
+    AddInput(
+        "Input",
+        "(Tensor)The input tensor, tensors with rank at most 6 are supported");
+    AddOutput("Output", "(Tensor)The output tensor");
```
Comment (Canpio) [inline paddle/operators/transpose_op.cc]: The names of a single in/out Op's input and output should be `X` and `Out` respectively. Here is the document: https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/name_convention.md

**Pull request merged**